### PR TITLE
Update the block component to use react hooks instead

### PR DESCRIPTION
**Pull request by youknowriad**
I said previously that we're not going to refactor our components to use hooks just to use hooks. But in this PR, I'm trying to refactor the `BlockListBlock` component. 

The main reasons are:

 - The component is one of the most complex one in the repository and using hooks clarifies its code as related behaviors/functions/state are grouped.
 - This is one of the bottleneck components in terms of performance, and I'd like to explore whether we can make improvements using hooks or not.
 - I have some plans to bring new features to this component and I don't want to just continue growing without some "organization".

At the moment, I kept the refactoring as minimal as I can, which means I kept, in theory, the same flows, states... There's something different though which can be impactful. A lot of the callbacks are not bound the instance object which means some child components might rerender more often. There are options and solutions but I want to measure the differences before making changes.

 - I can use `useCallback` to memoize these callbacks
 - I want to be mindful of the memory impact of memoizing all of these callbacks, maybe these are rerendered regardless of the callback change which means the memoization can be more costly than anticipated. 
 - Another option could be to move some of these callbacks to the `withDispatch` HoC instead.

### CODE REVIEW

**Comment by mtias** [comment]
This is a good start. Making this code more straightforward is going to help always keeping a cleaner state for any future iterations.

There's another motivation for making use of hooks here which is getting some more concrete awareness of how they can later benefit the consumable block API (exposing some of what has been so far absorbed in `support` or higher order components through hooks) and its tradeoffs.

**Comment by youknowriad** [comment]
I run the performance tests to compare this branch against master and noticed similar times. This refactor has no impact on performance. 

Can I get a review here, I'm thinking of merging this as is and try other smaller improvements using hooks separately.

File: packages/block-editor/src/components/block-list/block.js [hunk 1]
```diff
@@ -49,108 +49,181 @@ import Inserter from '../inserter';
...
+	onInsertBlocksAfter,
+	onMerge,
+	onSelect,
+	onRemove,
+	onInsertDefaultBlockAfter,
+	toggleSelection,
+	onShiftSelection,
+	onSelectionStart,
+} ) {
+	// Random state used to rerender the component if needed, ideally we don't need this
```
Comment (mtias) [inline packages/block-editor/src/components/block-list/block.js]: Why do we need it?

Comment (youknowriad) [reply packages/block-editor/src/components/block-list/block.js]: In two places, in the class implementation, we were calling `this.forceUpdate` to rerender. This doesn't exist in hooks and this is a workaround to allow it. I have plans to do more "hooks" refactorings to get rid of it separately.

File: packages/block-editor/src/components/block-list/block.js [hunk 2]
```diff
@@ -49,108 +49,181 @@ import Inserter from '../inserter';
...
+	const [ , updateRerenderState ] = useState( {} );
+	const rerender = () => updateRerenderState( {} );
+
+	// Reference of the wrapper -------------------------------------------------
```
Comment (aduth) [inline packages/block-editor/src/components/block-list/block.js]: What's with the dashes? I'd rather we embrace the pain of navigating "sections" of code to motivate us toward refactoring to separate functions/effects/etc, than to set some precedent for sectioning by code comments.

File: packages/block-editor/src/components/block-list/block.js [hunk 3]
```diff
@@ -49,108 +49,181 @@ import Inserter from '../inserter';
...
+	const wrapper = useRef( null );
+	useEffect( () => {
+		blockRef( wrapper.current, clientId );
+		// We need to rerender to trigger a rerendering of HoverArea.
+		rerender();
+	}, [] );
```
Comment (aduth) [inline packages/block-editor/src/components/block-list/block.js]: I see ESLint warnings here about needing to define `wrapper` and `clientId` as dependencies.

Aside: Is there a reason these ESLint rules are defined as warnings, not errors?

File: packages/block-editor/src/components/block-list/block.js [hunk 4]
```diff
@@ -316,309 +292,295 @@ export class BlockListBlock extends Component {
...
 
-	onBlockError( error ) {
-		this.setState( { error } );
-	}
+	/**
+	 * Begins tracking cursor multi-selection when clicking down within block.
+	 *
+	 * @param {MouseEvent} event A mousedown event.
+	 *
+	 * @return {void}
```
Comment (aduth) [inline packages/block-editor/src/components/block-list/block.js]: I think generally we're just not adding `@return` tag when the return value is `undefined`.

File: packages/block-editor/src/components/block-list/block.js [hunk 5]
```diff
@@ -49,108 +49,181 @@ import Inserter from '../inserter';
...
+	useEffect( () => {
+		blockRef( wrapper.current, clientId );
+		// We need to rerender to trigger a rerendering of HoverArea.
+		rerender();
+	}, [] );
+
+	// Reference to the block edit node -----------------------------------------
+	const blockNodeRef = useRef();
+
+	// Keep track of touchstart to disable hover on iOS -------------------------
```
Comment (aduth) [inline packages/block-editor/src/components/block-list/block.js]: Not for here, but I'm a bit curious if/why this hover tracking state is necessary, or whether there be a more elegant (abstracted effect?) way to handle it. The original discussions in #3139 and #2601 aren't entirely clear here.

**Review by aduth — APPROVED** [review]
It's a bit difficult to review, but it appears in good shape. I like how hooks help to consolidate where previously we have disjointedness of checking for changes in props values between `componentDidMount`, `componentDidUpdate`.

The issue with the ESLint warning should probably be resolved. There's also a merge conflict.

File: packages/block-editor/src/components/block-list/block.js (hunk 3 above)
Comment (youknowriad) [reply packages/block-editor/src/components/block-list/block.js]: After some hooks explorations in this PR and others, I think I'm going to remove this rule entirely as there are valid use-cases to omit dependencies (I want to only run this effect once). I feel that's why the React team suggests to use warnings here.

**Comment by hypest** [comment]
@Tug , can you give this PR a quick look to see if it can affect the native mobile in any way? cc @youknowriad

**Comment by Tug** [comment]
@hypest afaics it's only touching `BlockListBlock` which we're not using at the moment.
It might affect the way we port BlockHolder to gutenberg but it should not break anything right now.

File: packages/block-editor/src/components/block-list/block.js (hunk 3 above)
Comment (nerrad) [reply packages/block-editor/src/components/block-list/block.js]: I'm not sure your should remove the rule entirely but maybe just ignore the rule for the valid uses?  Hooks are still fairly new to the ecosystem and warnings can help contributors spot improper uses.

Comment (youknowriad) [reply packages/block-editor/src/components/block-list/block.js]: I did three refactorings to use hooks, in all of them, I found myself violating this rule. We can reconsider if there's an alternative way of using hooks, something I didn't discover yet for these use-cases which IMO are common.

Comment (nerrad) [reply packages/block-editor/src/components/block-list/block.js]: I'm still wary about usage of this pattern.  React Docs warn pretty strongly against it and Dan Abramov uses pretty strict warnings about it in an [article he dedicated to `useEffect`](https://overreacted.io/a-complete-guide-to-useeffect/):

> “But I only want to run it on mount!”, you’ll say. For now, remember: if you specify deps, all values from inside your component that are used by the effect must be there. Including props, state, functions — anything in your component.

Could it be written something like this?

```js
const wrapper = useRef( null );
const callBlockRef = () => blockRef( wrapper.current, clientId );
useEffect( () => {
    callBlockRef();
    // we need to rerender to trigger a rerendering of Hover Area.
    rerender();
} );
```

Comment (nerrad) [reply packages/block-editor/src/components/block-list/block.js]: Sorry my example is bad, something like this maybe:

```js
const callBlockRef = useCallback( ( wrapper, client.Id ) => {
    blockRef( wrapper.current, clientId );
}, [] );
useEffect( () => {
   callBlockRef();
  // we need to rerender to trigger a rerendering of Hover Area.
  rerender();
}, [ callBlockRef ] );
```

Comment (youknowriad) [reply packages/block-editor/src/components/block-list/block.js]: I agree this is not great code, but I think the fix for this is to not use `blockRef` prop entirely and just use ref forwarding to get the node from the block list. I tried doing this and I hit a wall where for some reason we're not ready yet to do it (I don't recall the exact reason as it was two weeks ago, probably related to BlockList not being hook ready)

File: packages/eslint-plugin/configs/react.js [hunk 6]
```diff
@@ -26,6 +26,5 @@ module.exports = {
 		'react/prop-types': 'off',
 		'react/react-in-jsx-scope': 'off',
 		'react-hooks/rules-of-hooks': 'error',
-		'react-hooks/exhaustive-deps': 'warn',
```
Comment (aduth) [inline packages/eslint-plugin/configs/react.js]: Can you explain why this was removed? Looking back through the history to https://github.com/WordPress/gutenberg/pull/14995#discussion_r275943117, I see there was some initial uncertainty about the rule. But as I've been reading more about it, it seems like one of those things where it should be a safe default to use, and that most problems end up from mis-use ([related discussion](https://github.com/facebook/react/issues/14920)). What are these valid exceptions we have, and are they really the majority of cases that we shouldn't only be opting-out as opposed to disabling this rule altogether?

Comment (youknowriad) [reply packages/eslint-plugin/configs/react.js]: When I first started trying hooks, it felt that we had a too many exceptions. Maybe related to the way our components were written or related to my thought process when writing components. happy to reconsider.

Comment (aduth) [reply packages/eslint-plugin/configs/react.js]: Truthfully, I'm still trying to understand all that the rule is meant to cover (its documentation... doesn't exist), but the general tone of the discussion around it from their end seems to be that if a developer is fighting with the rule, there's a higher likelihood the developer is "doing it wrong", so I'm curious to see a few specifics.

Maybe I'll flip the rule on locally to see what issues crop up, in order to get a better understanding.

Comment (epiqueras) [reply packages/eslint-plugin/configs/react.js]: It:

- Forces you to pass `deps = []`.
- Statically analyzes callbacks for closures and forces you to add them to `deps`.

**Pull request merged**